afamqp: add support for external authentication

spencercw opened the pull request:
AMQP has support for an 'external' authentication scheme which uses some out-of-band method for authenticating the user. For instance, the user may be identified from the Distinguished Name or Common Name in their client certificate. The method by which the user is identified when using external authentication depends on the server configuration.

Add a new 'auth-method' parameter to the amqp destination configuration which may be set to 'plain' or 'external'. When this parameter is omitted or set to 'plain', the existing behaviour is unchanged. When set to 'external' then external authentication is enabled and the username, if supplied, will be used as the identity, although this is generally not needed and may be omitted.

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
do nothing -> CI won't start)

Kokan commented:
@kira-syslogng ok to test

bazsi commented on modules/afamqp/afamqp.c:
```diff
@@ -463,8 +488,28 @@ afamqp_dd_connect(AMQPDestDriver *self, gboolean reconnect)
+                       self->auth_method, identity);
+      break;
+
+    default:
```
Since we validate when the grammar is parsed, this can easily become an assertion. No need to handle errors as they will essentially be dead code. Just use g_assert_not_reached ()

Kokan replied:
Travis compiles with `-Wswitch-default`, so it is kinda forced by us.

bazsi commented on modules/afamqp/afamqp.c:
```diff
@@ -463,8 +488,28 @@ afamqp_dd_connect(AMQPDestDriver *self, gboolean reconnect)
 
-  ret = amqp_login(self->conn, self->vhost, self->max_channel, self->frame_size, 0,
-                   AMQP_SASL_METHOD_PLAIN, self->user, self->password);
+  switch (self->auth_method)
```
Could you extract this login code to a function of it's own right? This connect function started to become to large.

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

spencercw commented:
External authentication is mentioned in the [RabbitMQ documentation](https://www.rabbitmq.com/authentication.html). I'm having trouble finding anything AMQP-specific so it's possible it's just a RabbitMQ thing.

I have addressed your comments. Thanks.
